### PR TITLE
Data-Diffing import

### PR DESCRIPTION
**Pull request by labkey-matthewb**
#### Rationale
Item 7784
Enable faster incremental import.  Current implementation does _not_ have a switch.  If DataIntegration module is present, we will attempt incremental import (list reimport, dataset import).

#### Related 
https://github.com/LabKey/platform/pull/1860
https://github.com/LabKey/dataintegration/pull/78
https://github.com/LabKey/labbook/pull/82

### CODE REVIEW

File: api/src/org/labkey/api/dataiterator/DataIterator.java [hunk 1]
```diff
@@ -33,8 +33,14 @@
  *
  *  Column 0 is the row number, used for error reporting
  */
-public interface DataIterator extends Closeable
+public interface DataIterator extends DataIteratorBuilder, Closeable
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/DataIterator.java:36]: :+1:

File: api/src/org/labkey/api/dataiterator/DataIteratorBuilder.java [hunk 2]
```diff
@@ -26,11 +26,10 @@
 {
     DataIterator getDataIterator(DataIteratorContext context);
 
+    /* Deprecated, finally made DataIterator simply extend DataIteratorBuilder */
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/DataIteratorBuilder.java]: Mark with `@Deprecated`?

File: api/src/org/labkey/api/dataiterator/DiskCachingDataIterator.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    int firstRowOnDisk = -1;    // index of first row written to disk
+    int lastRowOnDisk = -1;     // index of last row written (number or rows = lastRowOnDisk-firstRowOnDisk+1)
+
+    private void writeToDisk()
+    {
+        if (null == _buffer)
+        {
+            try
+            {
+            _tempFile = File.createTempFile("buffer", "dat");
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/DiskCachingDataIterator.java]: indentation

File: api/src/org/labkey/api/dataiterator/ListofMapsDataIterator.java [hunk 4]
```diff
@@ -51,11 +52,19 @@ public ListofMapsDataIterator(Set<String> colNames, List<Map<String, Object>> ro
     {
         super(null);
         _cols.add(new BaseColumnInfo("_rowNumber", JdbcType.INTEGER));
+        if (null == colNames || colNames.isEmpty())
+        {
+            CaseInsensitiveHashSet set = new CaseInsensitiveHashSet();
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/ListofMapsDataIterator.java]: I recently discovered that CaseInsensitiveHashSet doesn't preserve casing.  When checking whether column names are vocabulary propertyURIs we need the original casing.  You can use `Sets.newCaseInsensitiveHashSet()` to get a CaseInsensitiveHashMap backed set which does preserve casing.

File: api/src/org/labkey/api/di/DataIntegrationService.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    {
+        ServiceRegistry.get().registerService(DataIntegrationService.class, impl);
+    }
+
+    void registerStepProviders();
+    @Nullable Integer runTransformNow(Container c, User u, String transformId) throws PipelineJobException, NotFoundException;
+
+    RemoteConnection getRemoteConnection(String name, Container c, @Nullable Logger log);
+
+    /* Execute an efficient reimport operation Create a dataIterator based on target.getQueryUpdateService()
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/di/DataIntegrationService.java]:
```suggestion
    /**
     * Execute an efficient reimport operation Create a dataIterator based on target.getQueryUpdateService()
```

File: api/src/org/labkey/api/dataiterator/HashDataIterator.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    public static DataIterator wrap(DataIterator in, DataIteratorContext context)
+    {
+        // always wrap input data iterator because DataLoader may return HASH_COLUMN_NAME, but it might not be filled in.
+        return new HashDataIterator(in, context);
+    }
+
+    private HashDataIterator(DataIterator in, DataIteratorContext context)
+    {
+        super(in, context);
+        selectAll(Set.of(HASH_COLUMN_NAME));
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/HashDataIterator.java:52]: Can more than one hash data iterator be in the pipeline?  Should you throw if the input data iterator has a HASH_COLUMN_NAME column already?

File: api/src/org/labkey/api/dataiterator/HashDataIterator.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        for (int src=skipElementZero?1:0 ; src<map.size() ; src++)
+        {
+            if (src!=hashIndex)
+                row[dst++] = map.get(src);
+        }
+        String hashValue = calculateRowHash(row, 0, row.length);
+        map.set(hashIndex, hashValue);
+    }
+
+    /*
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/HashDataIterator.java:98]:
```suggestion
    /**
```

File: api/src/org/labkey/api/dataiterator/HashDataIterator.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        boolean ret = super.next();
+        if (ret)
+        {
+            if (null == _row[outputHashColumnIndex])
+                _row[outputHashColumnIndex] = calculateRowHash(_row,1,outputHashColumnIndex);
+        }
+        return ret;
+    }
+
+    /*
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/HashDataIterator.java]:
```suggestion
    /**
```

File: api/src/org/labkey/api/dataiterator/HashDataIterator.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+import org.labkey.api.data.BaseColumnInfo;
+import org.labkey.api.data.JdbcType;
+import org.labkey.api.di.DataIntegrationService;
+import org.labkey.api.query.BatchValidationException;
+import org.labkey.api.security.Crypt;
+
+import java.util.Set;
+import java.util.function.Supplier;
+
+/*
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/HashDataIterator.java]:
```suggestion
/**
```

File: api/src/org/labkey/api/di/DataIntegrationService.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        UPDATE, REPLACE, // pick ONE
+        DELETE
+    }
+
+
+    interface ReimportDataBuilder
+    {
+        void setReimportOptions(Set<ReimportOperations> ops);
+        ReimportDataBuilder setTarget(TableInfo tableInfo);
+        ReimportDataBuilder setSource(DataIteratorBuilder dib);
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/di/DataIntegrationService.java:80]: Should this be:
```suggestion
        ReimportDataBuilder setSource(ScrollableDataIterator dib);
```

File: api/src/org/labkey/api/dataiterator/DiskCachingDataIterator.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+        {
+            try
+            {
+                _index.add(_bufferEndPosition);
+                while (true)
+                {
+                    try
+                    {
+                        _buffer.position(_bufferEndPosition);
+                        ObjectOutputStream objectOutputStream = new ObjectOutputStream(new ByteBufferBackedOutputStream(_buffer));
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/dataiterator/DiskCachingDataIterator.java:141]: Given our experience with object streams, should we use a different serialization format?

File: api/src/org/labkey/api/di/DataIntegrationService.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        /* CONSDIER: extend TableInfo so it can supply this DataIterator (e.g. DataIterator getPrimaryKeyDataiterator(DataIterator in)) */
+        void setPrimaryKeyHelper(BiFunction<DataIterator, DataIteratorContext, DataIterator> wrap);
+
+        /* Ignore import data with modified <= modifiedSince.  This option is not supported with ReimportOperations.DELETE */
+        void setModifiedSince(Timestamp modifiedSince);
+
+        /** config params to pass to QUS insert/merge */
+        public void setConfigParameters(Map<Enum,Object> config);
+
+        void validate(BatchValidationException ex);
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/di/DataIntegrationService.java]: Can you add javadoc for this?  I initially thought it was validating the data, but it is validating the source and target can be used for reimporting.

**Review by labkey-kevink — APPROVED** [review]

File: api/src/org/labkey/api/di/DataIntegrationService.java (hunk 10 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/di/DataIntegrationService.java:80]: I think DataIteratorBuilder is right

File: api/src/org/labkey/api/dataiterator/DiskCachingDataIterator.java (hunk 11 above)
Comment (labkey-matthewb) [reply api/src/org/labkey/api/dataiterator/DiskCachingDataIterator.java:141]: This is fine since we have full control of the data.

**Pull request merged**